lib/pbench/agent/tool_group.py

```diff
@@ -0,0 +1,158 @@
+        except OSError as ex:
+            if ex.errno != errno.ENOENT:
+                raise
+            # Ignore missing trigger file
+            self.trigger = None
+        else:
+            if len(_trigger) == 0:
+                # Ignore empty trigger file contents
+                self.trigger = None
+            else:
+                self.trigger = _trigger
+
+        # toolnames - Dict with tool name as the key, dictionary with host
+        # names and parameters for each host
+        self.toolnames = {}
+        # hostnames - Dict with host name as the key, dictionary with tool
+        # names and parameters for each tool
+        self.hostnames = {}
+        self.labels = {}
+        for hdirent in os.listdir(self.tg_dir):
+            if hdirent == "__trigger__":
+                # Ignore handled above
+                continue
+            if not (self.tg_dir / hdirent).is_dir():
+                # Ignore wayward non-directory files
+                continue
+            # We assume this directory is a hostname.
+            host = hdirent
+            if host not in self.hostnames:
+                self.hostnames[host] = {}
+            for tdirent in os.listdir(self.tg_dir / host):
+                if tdirent == "__label__":
+                    self.labels[host] = (
+                        (self.tg_dir / host / tdirent).read_text().strip()
+                    )
+                    continue
+                if tdirent.endswith("__noinstall__"):
+                    # FIXME: ignore "noinstall" for now, tools are going to be
+                    # in containers so this does not make sense going forward.
+                    continue
+                # This directory entry is the name of a tool.
+                tool = tdirent
+                tool_opts_raw_lines = (
+                    (self.tg_dir / host / tool).read_text().split("\n")
+                )
+                tool_opts_lines = []
+                for line_raw in tool_opts_raw_lines:
+                    line = line_raw.strip()
+                    if not line:
+                        # Ignore blank lines
+                        continue
+                    tool_opts_lines.append(line)
+                tool_opts = " ".join(tool_opts_lines)
+                if tool not in self.toolnames:
+                    self.toolnames[tool] = {}
+                self.toolnames[tool][host] = tool_opts
+
+    def get_tools(self, host):
+        """get_tools - given a target host, return a dictionary with the list
+        of tool names as keys, and the values being their options for that
+        host.
+        """
+        tools = dict()
+        for tool, opts in self.toolnames.items():
+            try:
+                host_opts = opts[host]
+            except KeyError:
+                # This host does not have this tool registered, ignore.
+                pass
+            else:
+                tools[tool] = host_opts
```

Comment:
How about replacing that big, ugly `try` block with this?
```
            # If the host does not have this tool registered, then ignore it.
            if host in opts:
                tools[tool] = opts[host]
```

Reply:
https://stackoverflow.com/questions/7604636/better-to-try-something-and-catch-the-exception-or-test-if-its-possible-first

Reply:
Off the cuff, I would say that that SO article doesn't apply here, because having no tool registered isn't really an exceptional case (e.g., it's not an error, and the code provides for it as a part of normal operation).

However, from a performance standpoint, Python once again produces surprising (to me...) results.  In the non-exception case, my suggested code runs a few percent slower than the `try` block.  I would surmise that that is because it's doing the lookup twice (once for the `in` operation and once for the fetch); however, removing that redundancy using the following code:
```
     tmp = opts.get(host)
     if tmp is not None:
         tools[tool] = tmp
```
actually *slowed* the execution down by a significant amount!

And, in the case where the key is not present, while the `try` block is much more expensive than the others (as I expected...), the `in` operator is considerably faster than the `.get()` operation!

So, clearly there are some flaws in my understanding of how this stuff actually works.  :-(